File: packages/serve-sim/Sources/SimNetProxy/SimNetProxy.m
// Per-app NSURLSession proxy for network capture.

#import <Foundation/Foundation.h>
#import <objc/runtime.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>

static NSString *const kSimNetProxyHost = @"127.0.0.1";
static long simNetProxyPort = 0;

static void simnet_log(NSString *format, ...) {
    va_list args;
    va_start(args, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:args];
    va_end(args);
    NSLog(@"[simnetproxy] %@", message);
}

/**
 * HTTPS has no public iOS constants — literal keys CFNetwork expects;
 * without them only cleartext is proxied.
 */
static NSDictionary *SimNetProxyDictionary(void) {
    return @{
        (NSString *)kCFNetworkProxiesHTTPEnable : @YES,
        (NSString *)kCFNetworkProxiesHTTPProxy : kSimNetProxyHost,
        (NSString *)kCFNetworkProxiesHTTPPort : @(simNetProxyPort),
        @"HTTPSEnable" : @YES,
        @"HTTPSProxy" : kSimNetProxyHost,
        @"HTTPSPort" : @(simNetProxyPort),
    };
}

/** Port file dies with the proxy, so a stale launch finds nothing. */
static long SimNetProxyPortFromFile(const char *path) {
    FILE *handle = fopen(path, "r");
    if (handle == NULL) return 0;
    char buffer[32] = {0};
    size_t read = fread(buffer, 1, sizeof(buffer) - 1, handle);
    fclose(handle);
    return read > 0 ? atol(buffer) : 0;
}

static BOOL SimNetProxyPortIsListening(long port) {
    int fd = socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) return NO;
    struct sockaddr_in addr = {0};
    addr.sin_family = AF_INET;
    addr.sin_port = htons((uint16_t)port);
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    BOOL reachable = connect(fd, (struct sockaddr *)&addr, sizeof(addr)) == 0;
    close(fd);
    return reachable;
}

static void SimNetProxyApply(NSURLSessionConfiguration *configuration) {
    if (configuration == nil || simNetProxyPort <= 0) return;
    configuration.connectionProxyDictionary = SimNetProxyDictionary();
}

#pragma mark - Swizzles

@interface NSURLSessionConfiguration (SimNetProxy)
@end

@implementation NSURLSessionConfiguration (SimNetProxy)

+ (NSURLSessionConfiguration *)simnet_defaultSessionConfiguration {
    NSURLSessionConfiguration *configuration = [self simnet_defaultSessionConfiguration];
    SimNetProxyApply(configuration);
    return configuration;
}

+ (NSURLSessionConfiguration *)simnet_ephemeralSessionConfiguration {
    NSURLSessionConfiguration *configuration = [self simnet_ephemeralSessionConfiguration];
    SimNetProxyApply(configuration);
    return configuration;
}

+ (NSURLSessionConfiguration *)simnet_backgroundSessionConfigurationWithIdentifier:(NSString *)identifier {
    NSURLSessionConfiguration *configuration =
        [self simnet_backgroundSessionConfigurationWithIdentifier:identifier];
    SimNetProxyApply(configuration);
    return configuration;
}

@end

static BOOL SwizzleClassMethod(Class cls, SEL orig, SEL swiz) {
    Method o = class_getClassMethod(cls, orig);
    Method s = class_getClassMethod(cls, swiz);
    if (!o || !s) {
        simnet_log(@"swizzle FAILED: +[%@ %@]", NSStringFromClass(cls), NSStringFromSelector(orig));
        return NO;
    }
    method_exchangeImplementations(o, s);
    return YES;
}

/** Apple system apps (Safari, SpringBoard, …) break under MITM — AppSSO/AuthKit hang the main thread. */
static BOOL SimNetProxyShouldSkipProcess(void) {
    NSString *bundleId = NSBundle.mainBundle.bundleIdentifier;
    if (bundleId.length == 0) {
        // XPC helpers / no bundle — leave them alone.
        return YES;
    }
    if ([bundleId hasPrefix:@"com.apple."]) {
        return YES;
    }
    return NO;
}

#pragma mark - Entry point

__attribute__((constructor))
static void SimNetProxyInit(void) {
    @autoreleasepool {
        if (SimNetProxyShouldSkipProcess()) {
            return;
        }
        // Port file is the boot-capture path; bare port remains for a single-launch inject.
        const char *portFile = getenv("SIMNET_PROXY_PORT_FILE");
        const char *port = getenv("SIMNET_PROXY_PORT");
        long candidate = portFile != NULL ? SimNetProxyPortFromFile(portFile)
                                          : (port != NULL ? atol(port) : 0);
        if (candidate <= 0) {
            return;
        }
        if (!SimNetProxyPortIsListening(candidate)) {
            simnet_log(@"proxy %ld is not listening; leaving this app unproxied", candidate);
            return;
        }
        simNetProxyPort = candidate;

        Class cls = NSURLSessionConfiguration.class;
        BOOL ok = YES;
        ok &= SwizzleClassMethod(cls, @selector(defaultSessionConfiguration),
                                @selector(simnet_defaultSessionConfiguration));
        ok &= SwizzleClassMethod(cls, @selector(ephemeralSessionConfiguration),
                                @selector(simnet_ephemeralSessionConfiguration));
        ok &= SwizzleClassMethod(cls, @selector(backgroundSessionConfigurationWithIdentifier:),
                                @selector(simnet_backgroundSessionConfigurationWithIdentifier:));

        // `sharedSession` is built too early to swizzle.
        simnet_log(@"loaded into pid %d (%@), proxy %@:%ld, swizzles %@",
                   getpid(), NSBundle.mainBundle.bundleIdentifier ?: @"?", kSimNetProxyHost, simNetProxyPort,
                   ok ? @"installed" : @"INCOMPLETE");
    }
}

File: packages/serve-sim/Sources/SimNetProxy/SimNetProxy.c
#include <arpa/inet.h>
#include <dlfcn.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <objc/runtime.h>
#include <stdio.h>
#include <stdatomic.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <unistd.h>

#include "../ServeSimCapabilityLoader/startup-capability.h"

typedef id (*ConfigurationFactory)(id, SEL);
typedef id (*BackgroundFactory)(id, SEL, id);

static struct {
    Class (*getClass)(const char *);
    Class (*getMetaclass)(id);
    Class (*getSuperclass)(Class);
    Method *(*copyMethods)(Class, unsigned int *);
    SEL (*methodName)(Method);
    SEL (*selector)(const char *);
    IMP (*implementation)(Method);
    IMP (*replaceMethod)(Method, IMP);
    void (*send)(void);
} runtime;

static _Atomic(ConfigurationFactory) originalDefault;
static _Atomic(ConfigurationFactory) originalEphemeral;
static _Atomic(BackgroundFactory) originalBackground;
static long proxyPort;
static atomic_flag initialized = ATOMIC_FLAG_INIT;

static Method findFactory(Class cls, const char *name) {
    SEL selector = runtime.selector(name);
    for (Class meta = runtime.getMetaclass((id)cls); meta; meta = runtime.getSuperclass(meta)) {
        unsigned int count = 0;
        Method *methods = runtime.copyMethods(meta, &count);
        Method found = NULL;
        for (unsigned int i = 0; i < count; i++) {
            if (runtime.methodName(methods[i]) == selector) {
                found = methods[i];
                break;
            }
        }
        free(methods);
        if (found) return found;
    }
    return NULL;
}

static id string(const char *value) {
    return ((id (*)(id, SEL, const char *))runtime.send)(
        (id)runtime.getClass("NSString"), runtime.selector("stringWithUTF8String:"), value);
}

static id number(long value) {
    return ((id (*)(id, SEL, long))runtime.send)(
        (id)runtime.getClass("NSNumber"), runtime.selector("numberWithLong:"), value);
}

static id applyProxy(id configuration) {
    if (!configuration) return configuration;
    id host = string("127.0.0.1");
    id port = number(proxyPort);
    id enabled = number(1);
    id keys[] = {string("HTTPEnable"), string("HTTPProxy"), string("HTTPPort"),
                 string("HTTPSEnable"), string("HTTPSProxy"), string("HTTPSPort")};
    id values[] = {enabled, host, port, enabled, host, port};
    id dictionary = ((id (*)(id, SEL, const id *, const id *, unsigned long))runtime.send)(
        (id)runtime.getClass("NSDictionary"), runtime.selector("dictionaryWithObjects:forKeys:count:"),
        values, keys, 6);
    ((void (*)(id, SEL, id))runtime.send)(
        configuration, runtime.selector("setConnectionProxyDictionary:"), dictionary);
    return configuration;
}

static id defaultConfiguration(id cls, SEL selector) {
    return applyProxy(atomic_load(&originalDefault)(cls, selector));
}

static id ephemeralConfiguration(id cls, SEL selector) {
    return applyProxy(atomic_load(&originalEphemeral)(cls, selector));
}

static id backgroundConfiguration(id cls, SEL selector, id identifier) {
    return applyProxy(atomic_load(&originalBackground)(cls, selector, identifier));
}

static long readPort(void) {
    const char *path = getenv("SIMNET_PROXY_PORT_FILE");
    if (!path) return 0;
    int file = open(path, O_RDONLY | O_NONBLOCK | O_NOFOLLOW);
    if (file < 0) return 0;
    struct stat info;
    if (fstat(file, &info) != 0 || !S_ISREG(info.st_mode)) {
        close(file);
        return 0;
    }
    char text[32] = {0};
    ssize_t length = read(file, text, sizeof(text) - 1);
    close(file);
    char *end;
    long port = strtol(text, &end, 10);
    if (length <= 0 || end == text || (*end != '\0' && *end != '\n') || port <= 0 || port > 65535) return 0;
    int socketFd = socket(AF_INET, SOCK_STREAM, 0);
    if (socketFd < 0) return 0;
    struct sockaddr_in address = {0};
    address.sin_family = AF_INET;
    address.sin_port = htons((uint16_t)port);
    address.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    int result = connect(socketFd, (struct sockaddr *)&address, sizeof address);
    close(socketFd);
    return result == 0 ? port : 0;
}

static void initializeProxy(void) {
    if (atomic_flag_test_and_set(&initialized)) return;
    proxyPort = readPort();
    if (!proxyPort) return;
    runtime.getClass = dlsym(RTLD_DEFAULT, "objc_getClass");
    runtime.getMetaclass = dlsym(RTLD_DEFAULT, "object_getClass");
    runtime.getSuperclass = dlsym(RTLD_DEFAULT, "class_getSuperclass");
    runtime.copyMethods = dlsym(RTLD_DEFAULT, "class_copyMethodList");
    runtime.methodName = dlsym(RTLD_DEFAULT, "method_getName");
    runtime.selector = dlsym(RTLD_DEFAULT, "sel_registerName");
    runtime.implementation = dlsym(RTLD_DEFAULT, "method_getImplementation");
    runtime.replaceMethod = dlsym(RTLD_DEFAULT, "method_setImplementation");
    runtime.send = dlsym(RTLD_DEFAULT, "objc_msgSend");
    if (!runtime.getClass || !runtime.getMetaclass || !runtime.getSuperclass || !runtime.copyMethods ||
        !runtime.methodName || !runtime.selector || !runtime.implementation || !runtime.replaceMethod || !runtime.send) return;
    Class cls = runtime.getClass("NSURLSessionConfiguration");
    if (!cls) return;
    Method defaults = findFactory(cls, "defaultSessionConfiguration");
    Method ephemeral = findFactory(cls, "ephemeralSessionConfiguration");
    Method background = findFactory(cls, "backgroundSessionConfigurationWithIdentifier:");
    if (!defaults || !ephemeral || !background) {
        fprintf(stderr, "[simnetproxy] session factories unavailable; relaunch with a supported simulator runtime.\n");
        return;
    }
    atomic_store(&originalDefault, (ConfigurationFactory)runtime.implementation(defaults));
    atomic_store(&originalEphemeral, (ConfigurationFactory)runtime.implementation(ephemeral));
    atomic_store(&originalBackground, (BackgroundFactory)runtime.implementation(background));
    runtime.replaceMethod(defaults, (IMP)defaultConfiguration);
    runtime.replaceMethod(ephemeral, (IMP)ephemeralConfiguration);
    runtime.replaceMethod(background, (IMP)backgroundConfiguration);
}

SERVE_SIM_STARTUP_CAPABILITY(initializeProxy)
